### ISSUE TITLE
Fulltext search returns no results when search includes certain special characters

### ISSUE THREAD

**Issue by briansandall** [cubecart/v6#1624]
This was mentioned in #1609 - product descriptions (both short and long) get fully HTML-encoded _prior_ to storing in the database, meaning characters such as the ampersand `&` will not return any hits when used in a search because the database instead has `&amp;`.

1. Create a product with any name and product code you like, but it must be different from whatever text you use for step 2

2. Put anything with an ampersand in the product description, e.g. 'M&M'

3. Search the store for the term in step 2 - no results

4. Search the store for the term in step 2 but replace `&` with `&amp;`, e.g. 'M&amp;M' - now you will get the expected products in the search results

For data integrity reasons, my preferred solution would be to change the code responsible for saving the description text to leave it completely unaltered. Note that the product name and product code already do not encode the ampersand, for example, so it is not unprecedented.

If an admin/dev ever needs to perform queries directly on the database, it is much easier if they do not need to remember that some fields have some characters encoded for HTML.

If that is not possible due to e.g. 3rd-party code, then a possible solution would be to apply the same logic to the search terms for the part of the query that does a fulltext search; since the other fields are not encoded, however, we still need the unencoded search terms for those parts of the query.

**Comment by abrookbanks** [comment]
I'm not entirely sure what the solution is. From issue #1624 I think this issue is resolved?

I wonder what the statistics are for searches with special characters. I'd have thought that a search would be more likely to be something along the lines of typical words or product codes. 

I'm not sure if I can do anything with this.

**Comment by briansandall** [comment]
Apostrophes work, yes, from #1609.

Ampersands, however, do not*. There may be other special characters as well, but everything I've tested (other than the ampersand) has worked fine.

I agree it is probably a small percentage of searches and may not be worth fixing at this time, but it is something that until fixed store owners should be made aware of so they don't rely on adding terms such as `M&M's` to their `M&M Candy` product description in hopes of enabling the product to be found via common search terms.

Again, probably not a common issue, but it is an issue. ;)

* They do work in a product name or product code; they do NOT work in a product description.

**Comment by abrookbanks** [comment]
Ok. Thanks for such a thorough explanation, thats very helpful. 

I'm not sure I can fix this minor issue.

**Comment by briansandall** [comment]
Fair enough.

The only solution I can think of that would be half-way reasonable would be to encode ampersands in the product names and product codes and then also in the search terms.

This is already done with apostrophes in all those cases, but ampersand is only encoded in the product descriptions.

But as you say, it may not be worth fixing at this time.

**Comment by Dirty-Butter** [comment]
We have lots of toy manufacturer brand names with & or with apostrophes.

**Comment by briansandall** [comment]
> We have lots of toy manufacturer brand names with & or with apostrophes.

Luckily, only product descriptions (long and short) suffer from the issue with the ampersand - the ampersand is stored as is everywhere else in the database (except for maybe email templates and documents). Searching by manufacturer also works normally.

However, if you include the manufacturer name in the product description and the customer does a _normal_ (i.e. not advanced) search for the manufacturer, your product will not come up if the manufacturer name includes an ampersand.

Apostrophes work properly now after #1609 was resolved.

**Comment by briansandall** [comment]
I've posted a potential fix - if anyone else has time to try it out, please do and let us know how it goes!

**Comment by briansandall** [comment]
Alright, seems to work as expected now.

**Comment by abrookbanks** [comment]
I need to buck up my game as this was actually a lot simpler than I thought. 

Pull request merged. Thanks for this.
